File: php/src/Metrics/NdjsonWriter.php
<?php

declare(strict_types=1);

namespace RespBench\Metrics;

use RuntimeException;

/**
 * Writes benchmark metrics to NDJSON (newline-delimited JSON), one line per phase.
 * The schema matches the Java/Ruby/Python engines exactly.
 */
final class NdjsonWriter
{
    private ?string $commitId = null;
    private ?string $driverId = null;
    private ?string $primaryDriverVersion = null;
    private ?string $secondaryDriverId = null;
    private ?string $secondaryDriverVersion = null;

    public function __construct(private readonly string $outputPath)
    {
    }

    public function setMetadata(
        ?string $commitId,
        ?string $driverId,
        ?string $primaryDriverVersion,
        ?string $secondaryDriverId = null,
        ?string $secondaryDriverVersion = null,
    ): void {
        $this->commitId = $commitId;
        $this->driverId = $driverId;
        $this->primaryDriverVersion = $primaryDriverVersion;
        $this->secondaryDriverId = $secondaryDriverId;
        $this->secondaryDriverVersion = $secondaryDriverVersion;
    }

    public function writePhaseResults(string $phaseId, string $status, int $connections, Collector $collector): void
    {
        $dir = dirname($this->outputPath);
        if (!is_dir($dir) && !mkdir($dir, 0o777, true) && !is_dir($dir)) {
            throw new RuntimeException("Cannot create output directory: {$dir}");
        }

        $json = $this->buildPhaseJson($phaseId, $status, $connections, $collector);
        $line = json_encode($json, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES);

        file_put_contents($this->outputPath, $line . "\n", FILE_APPEND | LOCK_EX);
    }

    /**
     * @return array<string,mixed>
     */
    private function buildPhaseJson(string $phaseId, string $status, int $connections, Collector $collector): array
    {
        $result = [];

        if ($this->commitId !== null || $this->driverId !== null) {
            $metadata = [];
            if ($this->commitId !== null) {
                $metadata['commit_id'] = $this->commitId;
            }
            $metadata['timestamp'] = gmdate('Y-m-d\TH:i:s\Z');
            if ($this->driverId !== null) {
                $metadata['driver_id'] = $this->driverId;
            }
            if ($this->primaryDriverVersion !== null) {
                $metadata['primary_driver_version'] = $this->primaryDriverVersion;
            }
            if ($this->secondaryDriverId !== null) {
                $metadata['secondary_driver_id'] = $this->secondaryDriverId;
            }
            if ($this->secondaryDriverVersion !== null) {
                $metadata['secondary_driver_version'] = $this->secondaryDriverVersion;
            }
            $result['metadata'] = $metadata;
        }

        $result['phase'] = [
            'id' => $phaseId,
            'status' => $status,
            'start_timestamp' => self::iso8601($collector->startTime()),
            'finish_timestamp' => self::iso8601($collector->endTime()),
            'duration_ms' => $collector->durationMillis(),
            'connections' => $connections,
        ];

        $result['totals'] = [
            'requests' => $collector->totalRequests(),
            'errors' => $collector->totalErrors(),
        ];

        $result['metrics'] = $this->buildCommandMetrics($collector);

        return $result;
    }

    /**
     * @return array<string,mixed>
     */
    private function buildCommandMetrics(Collector $collector): array
    {
        $metrics = [];

        foreach ($collector->allMetrics() as $name => $cmd) {
            $data = [
                'requests' => $cmd->requests(),
                'errors' => $cmd->errors(),
                'latency' => [
                    'unit' => 'us',
                    'count' => $cmd->count(),
                    'summary' => [
                        'min' => $cmd->min(),
                        'p50' => $cmd->percentile(50),
                        'p95' => $cmd->percentile(95),
                        'p99' => $cmd->percentile(99),
                        'p999' => $cmd->percentile(99.9),
                        'max' => $cmd->max(),
                    ],
                ],
            ];

            $histogram = $cmd->histogram();
            if ($histogram !== null) {
                $data['latency']['hdr'] = [
                    'format' => 'hdr',
                    'sigfig' => 3,
                    'payload_b64' => self::encodeHistogram($histogram),
                ];
            }

            $metrics[$name] = $data;
        }

        return $metrics;
    }

    private static function encodeHistogram(HdrHistogram $histogram): string
    {
        try {
            return HdrEncoder::encodeCompressedBase64($histogram);
        } catch (\Throwable) {
            return '';
        }
    }

    private static function iso8601(?float $time): ?string
    {
        if ($time === null) {
            return null;
        }

        return gmdate('Y-m-d\TH:i:s\Z', (int) $time);
    }
}

File: go/internal/metrics/ndjsonwriter.go
package metrics

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// NdjsonWriter appends one JSON line per phase, matching the schema used by the
// Java/Ruby/Python/PHP/Node engines exactly.
type NdjsonWriter struct {
	path string
	mu   sync.Mutex

	commitID               string
	driverID               string
	primaryDriverVersion   string
	secondaryDriverID      string
	secondaryDriverVersion string
	driverDetails          map[string]any
	hasMetadata            bool
}

// NewNdjsonWriter creates a writer for the given output path.
func NewNdjsonWriter(path string) *NdjsonWriter {
	return &NdjsonWriter{path: path}
}

// SetMetadata records the run metadata written into every phase line. Extra
// driverDetails keys (negotiated protocol, parser, pipelining mode) are merged
// into the metadata block; downstream tooling reads metadata by key, so extra
// keys are harmless to engines that omit them.
func (w *NdjsonWriter) SetMetadata(commitID, driverID, primaryVersion, secondaryID, secondaryVersion string, driverDetails map[string]any) {
	w.commitID = commitID
	w.driverID = driverID
	w.primaryDriverVersion = primaryVersion
	w.secondaryDriverID = secondaryID
	w.secondaryDriverVersion = secondaryVersion
	w.driverDetails = driverDetails
	w.hasMetadata = true
}

// summary is the latency percentile summary block.
type summary struct {
	Min  int64 `json:"min"`
	P50  int64 `json:"p50"`
	P95  int64 `json:"p95"`
	P99  int64 `json:"p99"`
	P999 int64 `json:"p999"`
	Max  int64 `json:"max"`
}

type hdrBlock struct {
	Format     string `json:"format"`
	Sigfig     int    `json:"sigfig"`
	PayloadB64 string `json:"payload_b64"`
}

type latencyBlock struct {
	Unit    string    `json:"unit"`
	Count   int64     `json:"count"`
	Summary summary   `json:"summary"`
	HDR     *hdrBlock `json:"hdr,omitempty"`
}

type commandBlock struct {
	Requests int64        `json:"requests"`
	Errors   int64        `json:"errors"`
	Latency  latencyBlock `json:"latency"`
}

type phaseBlock struct {
	ID              string  `json:"id"`
	Status          string  `json:"status"`
	StartTimestamp  *string `json:"start_timestamp"`
	FinishTimestamp *string `json:"finish_timestamp"`
	DurationMs      int64   `json:"duration_ms"`
	Connections     int     `json:"connections"`
	// Additive, optional fields. `connections` alone cannot distinguish a
	// pipelined run from a serial one, and the concurrency actually applied is
	// connections x pipeline_depth. sockets_per_client is 1 for a multiplexing
	// driver at any depth, pipeline_depth for a pooling one. Other engines omit
	// these and downstream tooling reads by key, so they are ignored there.
	PipelineDepth    int `json:"pipeline_depth"`
	SocketsPerClient int `json:"sockets_per_client"`
	TotalSockets     int `json:"total_sockets"`
}

// phaseRecord is the full per-phase JSON object. Metadata is a free-form map so
// driver-detail keys can be merged alongside the fixed keys. Metrics is always a
// JSON object ({} when empty), never an array, because consumers index into it.
type phaseRecord struct {
	Metadata map[string]any          `json:"metadata,omitempty"`
	Phase    phaseBlock              `json:"phase"`
	Totals   map[string]int64        `json:"totals"`
	Metrics  map[string]commandBlock `json:"metrics"`
}

// WritePhaseResults appends a phase record built from the collector.
func (w *NdjsonWriter) WritePhaseResults(phaseID, status string, connections, pipelineDepth, socketsPerClient int, c *Collector) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if err := os.MkdirAll(filepath.Dir(w.path), 0o755); err != nil {
		return fmt.Errorf("create output dir: %w", err)
	}

	rec := w.buildRecord(phaseID, status, connections, pipelineDepth, socketsPerClient, c)
	line, err := json.Marshal(rec)
	if err != nil {
		return fmt.Errorf("marshal phase record: %w", err)
	}

	f, err := os.OpenFile(w.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("open metrics file: %w", err)
	}
	defer f.Close()
	if _, err := f.Write(append(line, '\n')); err != nil {
		return fmt.Errorf("write metrics line: %w", err)
	}
	return nil
}

func (w *NdjsonWriter) buildRecord(phaseID, status string, connections, pipelineDepth, socketsPerClient int, c *Collector) phaseRecord {
	if socketsPerClient < 1 {
		socketsPerClient = 1
	}
	rec := phaseRecord{
		Phase: phaseBlock{
			ID:               phaseID,
			Status:           status,
			StartTimestamp:   iso8601(c.StartTime()),
			FinishTimestamp:  iso8601(c.StopTime()),
			DurationMs:       c.DurationMillis(),
			Connections:      connections,
			PipelineDepth:    pipelineDepth,
			SocketsPerClient: socketsPerClient,
			TotalSockets:     connections * socketsPerClient,
		},
		Totals: map[string]int64{
			"requests": c.TotalRequests(),
			"errors":   c.TotalErrors(),
		},
		Metrics: map[string]commandBlock{}, // marshals as {} when empty
	}

	if w.hasMetadata {
		md := map[string]any{
			"timestamp": time.Now().UTC().Format("2006-01-02T15:04:05Z"),
		}
		if w.commitID != "" {
			md["commit_id"] = w.commitID
		}
		if w.driverID != "" {
			md["driver_id"] = w.driverID
		}
		if w.primaryDriverVersion != "" {
			md["primary_driver_version"] = w.primaryDriverVersion
		}
		if w.secondaryDriverID != "" {
			md["secondary_driver_id"] = w.secondaryDriverID
		}
		if w.secondaryDriverVersion != "" {
			md["secondary_driver_version"] = w.secondaryDriverVersion
		}
		// Merge driver-detail keys without overwriting the fixed keys.
		for k, v := range w.driverDetails {
			if _, exists := md[k]; !exists {
				md[k] = v
			}
		}
		rec.Metadata = md
	}

	for name, cm := range c.Commands() {
		block := commandBlock{
			Requests: cm.Requests,
			Errors:   cm.Errors,
			Latency: latencyBlock{
				Unit:  "us",
				Count: cm.Histogram.TotalCount(),
				Summary: summary{
					Min:  cm.Histogram.Min(),
					P50:  cm.Histogram.ValueAtPercentile(50),
					P95:  cm.Histogram.ValueAtPercentile(95),
					P99:  cm.Histogram.ValueAtPercentile(99),
					P999: cm.Histogram.ValueAtPercentile(99.9),
					Max:  cm.Histogram.Max(),
				},
			},
		}
		if b64, err := EncodeCompressedBase64(cm.Histogram); err == nil {
			block.Latency.HDR = &hdrBlock{Format: "hdr", Sigfig: 3, PayloadB64: b64}
		}
		rec.Metrics[name] = block
	}
	return rec
}

// iso8601 formats a UTC timestamp, or nil if the time is zero.
func iso8601(t time.Time) *string {
	if t.IsZero() {
		return nil
	}
	s := t.UTC().Format("2006-01-02T15:04:05Z")
	return &s
}
